### PR TITLE
Verify managed artifact digests before activation

### PR DESCRIPTION
**Pull request by Zigfreidish**
## Summary
- Refs #1258.
- Add strict managed direct-download SHA-256 verification against staged `*.partial` bytes before artifact activation.
- Record declared and actual digests, real integrity timestamps, and fail-closed receipts for digest mismatch or unsupported digest policy.
- Add strict managed Hub repository snapshot SHA-256 verification over deterministic directory contents before activation.
- Expose normalized download diagnostics rows with `artifact_integrity_status` and `artifact_integrity_summary` fields for operation receipts.
- Update the Phase 8 local install runbook and add focused implementation plans for these slices.

## Plan or Spec
- `docs/plans/2026-06-21-issue-1258-digest-verification.md`
- `docs/plans/2026-06-22-issue-1258-hub-snapshot-digest.md`
- `docs/plans/2026-06-22-issue-1258-diagnostics-integrity-fields.md`
- `docs/runbooks/phase-8-local-install.md`

## Commands Run
```text
make bootstrap
# passed: configured core.hooksPath=.githooks; uv sync resolved 122 packages and checked 79 packages.

make proto
# passed; git status remained clean.

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python pytest -q services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_managed_download_with_digest_materializes_artifact services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_managed_download_rejects_digest_mismatch_before_activation services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_artifact_integrity_receipt_normalizes_bare_sha256_digest services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_managed_download_rejects_unsupported_digest_policy_before_activation
# 4 passed in 0.04s after merging origin/main

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python pytest -q services/mlx-worker-python/tests/test_download_pipeline_unit.py services/mlx-worker-python/tests/test_model_ops_job_registry.py services/mlx-worker-python/tests/test_managed_artifact_receipts.py
# 69 passed in 0.50s after merging origin/main

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python coverage run -m pytest -q services/mlx-worker-python/tests/test_download_pipeline_unit.py services/mlx-worker-python/tests/test_model_ops_job_registry.py services/mlx-worker-python/tests/test_managed_artifact_receipts.py
PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python coverage json -o coverage.json
python3 scripts/changed_scope_coverage.py --coverage-json coverage.json services/mlx-worker-python/worker/model_ops/download_pipeline.py services/mlx-worker-python/tests/test_download_pipeline_unit.py
# TOTAL 143 3 98%

python3 -m py_compile services/mlx-worker-python/worker/model_ops/download_pipeline.py services/mlx-worker-python/tests/test_download_pipeline_unit.py
# passed

git diff --check
# passed

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python python scripts/pr_scoped_performance_run.py --registry infra/perf/pr_scoped_probes.json --probe-id download-pipeline-directory-size-single-stat --base-repo .runtime/pr/perf-merge-check/base --head-repo .runtime/pr/perf-merge-check/head --coverage-paths-json '["services/mlx-worker-python/tests/test_download_pipeline_unit.py","services/mlx-worker-python/worker/model_ops/download_pipeline.py"]' --output .runtime/pr/perf-merge-check/probes/download-pipeline-directory-size-single-stat.json
PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python python scripts/pr_scoped_performance_report.py --scope .runtime/pr/perf-merge-check/scope/scope.json --results-dir .runtime/pr/perf-merge-check/probes --format terminal --output-dir .runtime/pr/perf-merge-check/report
# Status: ok; selected probes 1; regressions 0.

git commit -m "Verify managed artifact digests before activation"
# pre-commit ran make swift-test, make py-test, make integration-test, and scoped performance successfully.
# make swift-test: completed rc=0 elapsed=698.6s
# make py-test: 4142 passed, 14 skipped, 2 warnings in 204.74s
# make integration-test: 122 passed, 1 skipped in 775.91s

git merge --no-edit origin/main
# merged origin/main at dc7686c9ae7a23178f450648ed939bc137ebf65b cleanly.

make swift-test
# passed after merging origin/main; final macOS menubar stage: 831 tests in 25 suites passed after 32.670s.

make py-test
# 4143 passed, 14 skipped, 2 warnings in 175.80s.

make integration-test
# 122 passed, 1 skipped in 695.02s.

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python pytest -q services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_integrity_rejects_missing_actual_digest_before_activation
# failed as expected before the review fix: Failed: DID NOT RAISE <class 'worker.model_ops.errors.ModelOperationError'>

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python pytest -q services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_managed_download_with_digest_materializes_artifact services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_managed_download_rejects_digest_mismatch_before_activation services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_integrity_rejects_missing_actual_digest_before_activation services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_artifact_integrity_receipt_normalizes_bare_sha256_digest services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_managed_download_rejects_unsupported_digest_policy_before_activation
# 5 passed in 0.10s.

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python pytest -q services/mlx-worker-python/tests/test_download_pipeline_unit.py services/mlx-worker-python/tests/test_model_ops_job_registry.py services/mlx-worker-python/tests/test_managed_artifact_receipts.py
# 70 passed in 0.59s.

python3 -m py_compile services/mlx-worker-python/worker/model_ops/download_pipeline.py services/mlx-worker-python/tests/test_download_pipeline_unit.py
# passed

git diff --check
# passed

git commit -m "Fail closed on missing artifact actual digest"
# pre-commit ran make swift-test, make py-test, make integration-test, and scoped performance successfully.
# make swift-test: completed rc=0 elapsed=452.8s; final macOS menubar stage: 831 tests in 25 suites passed after 30.920s.
# make py-test: 4144 passed, 14 skipped, 2 warnings in 170.92s.
# make integration-test: 122 passed, 1 skipped in 643.72s.

git merge --no-edit origin/main
# merged origin/main at e2014b2aa9f9c4fc7d00a4679f25ec1b5671d629 cleanly.

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python pytest -q services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_managed_download_with_digest_materializes_artifact services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_managed_download_rejects_digest_mismatch_before_activation services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_integrity_rejects_missing_actual_digest_before_activation services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_artifact_integrity_receipt_normalizes_bare_sha256_digest services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_managed_download_rejects_unsupported_digest_policy_before_activation
# 5 passed in 0.11s after merging latest origin/main.

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python pytest -q services/mlx-worker-python/tests/test_download_pipeline_unit.py services/mlx-worker-python/tests/test_model_ops_job_registry.py services/mlx-worker-python/tests/test_managed_artifact_receipts.py
# 70 passed in 1.99s after merging latest origin/main.

python3 -m py_compile services/mlx-worker-python/worker/model_ops/download_pipeline.py services/mlx-worker-python/tests/test_download_pipeline_unit.py
# passed after merging latest origin/main.

git diff --check
# passed after merging latest origin/main.

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python python3 - <<'PY'
from pathlib import Path
from scripts.pre_commit_gate import run_performance_report

changed_files = [
    "docs/plans/2026-06-21-issue-1258-digest-verification.md",
    "docs/runbooks/phase-8-local-install.md",
    "services/mlx-worker-python/tests/test_download_pipeline_unit.py",
    "services/mlx-worker-python/worker/model_ops/download_pipeline.py",
]
outcome = run_performance_report(Path.cwd(), changed_files, base_ref="origin/main")
print(f"report_dir={outcome.report_dir}")
print(f"status={outcome.status}")
print(f"selected_probe_count={outcome.selected_probe_count}")
PY
# Status: ok; selected probes 1; regressions 0; coverage 98.0% after merging latest origin/main.

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python pytest -q services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_managed_hub_import_with_matching_snapshot_digest_activates services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_managed_hub_import_rejects_snapshot_digest_mismatch_before_activation services/mlx-worker-python/tests/test_managed_artifact_receipts.py::test_strict_managed_hub_download_failure_emits_snapshot_digest_receipt services/mlx-worker-python/tests/test_maintenance_service.py::test_download_job_reports_managed_hub_snapshot_without_creating_descriptor services/mlx-worker-python/tests/test_maintenance_service.py::test_managed_hub_repo_download_uses_default_huggingface_cache_and_cached_token services/mlx-worker-python/tests/test_maintenance_service.py::test_download_job_marks_managed_dflash_draft_metadata
# 6 passed in 0.32s.

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python pytest -q services/mlx-worker-python/tests/test_download_pipeline_unit.py services/mlx-worker-python/tests/test_managed_artifact_receipts.py
PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python pytest -q services/mlx-worker-python/tests/test_maintenance_service.py::test_download_job_reports_managed_hub_snapshot_without_creating_descriptor services/mlx-worker-python/tests/test_maintenance_service.py::test_managed_hub_repo_download_uses_default_huggingface_cache_and_cached_token services/mlx-worker-python/tests/test_maintenance_service.py::test_download_job_marks_managed_dflash_draft_metadata
python3 -m py_compile services/mlx-worker-python/worker/model_ops/download_pipeline.py services/mlx-worker-python/tests/test_download_pipeline_unit.py services/mlx-worker-python/tests/test_managed_artifact_receipts.py
git diff --check
# 37 passed in 0.53s; 3 passed in 0.29s; py_compile passed; git diff --check passed.

rm -f coverage.json
PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python coverage run -m pytest -q services/mlx-worker-python/tests/test_download_pipeline_unit.py services/mlx-worker-python/tests/test_managed_artifact_receipts.py
PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python coverage json -o coverage.json
python3 scripts/changed_scope_coverage.py --coverage-json coverage.json services/mlx-worker-python/worker/model_ops/download_pipeline.py services/mlx-worker-python/tests/test_download_pipeline_unit.py services/mlx-worker-python/tests/test_managed_artifact_receipts.py
rm -f coverage.json
# TOTAL 128 0 100%.

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python python3 scripts/pr_scoped_performance_scope.py --registry infra/perf/pr_scoped_probes.json --changed-files-json .runtime/pr/issue-1258-hub-snapshot/scope/changed-files.json --output .runtime/pr/issue-1258-hub-snapshot/scope/scope.json
PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python python3 scripts/pr_scoped_performance_run.py --registry infra/perf/pr_scoped_probes.json --probe-id download-pipeline-directory-size-single-stat --base-repo .runtime/pr/perf-merge-check/base --head-repo . --coverage-paths-json '["services/mlx-worker-python/tests/test_download_pipeline_unit.py","services/mlx-worker-python/worker/model_ops/download_pipeline.py"]' --output .runtime/pr/issue-1258-hub-snapshot/probes/download-pipeline-directory-size-single-stat.json
PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python python3 scripts/pr_scoped_performance_report.py --scope .runtime/pr/issue-1258-hub-snapshot/scope/scope.json --results-dir .runtime/pr/issue-1258-hub-snapshot/probes --format terminal --output-dir .runtime/pr/issue-1258-hub-snapshot/report
# Status: ok; selected probes 1; regressions 0; coverage 100.0%.
# elapsed_ms_mean base 37.578 ms, head 36.986 ms, delta -1.57%; elapsed_ms_min base 36.807 ms, head 36.497 ms, delta -0.84%.

git commit -m "Verify managed hub snapshot digests"
# pre-commit ran make swift-test, make py-test, make integration-test, and scoped performance successfully.
# make swift-test: completed rc=0 elapsed=205.6s.
# make py-test: 4150 passed, 14 skipped, 2 warnings in 169.74s.
# make integration-test: 122 passed, 1 skipped in 424.66s.
# pre-commit scoped performance: status ok; selected probes 1; regressions 0; coverage 100.0%.
# pre-commit performance report: .runtime/pre-commit-performance/20260621-162228-d2017dc4/report/report.md.
# pre-commit metrics: elapsed_ms_mean base 37.198 ms, head 36.711 ms, delta -1.31%; elapsed_ms_min base 36.715 ms, head 36.155 ms, delta -1.53%.

git fetch origin main:refs/remotes/origin/main
git rev-parse --short origin/main
git merge-base --is-ancestor origin/main HEAD
# origin/main e2014b2a; origin-main-is-ancestor=yes.

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python pytest -q services/mlx-worker-python/tests/test_model_ops_job_registry.py::test_download_registry_snapshot_exposes_operation_receipt_fields
# failed as expected before the diagnostics fix: KeyError: 'artifact_integrity_summary'

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python pytest -q services/mlx-worker-python/tests/test_model_ops_job_registry.py::test_download_registry_snapshot_exposes_operation_receipt_fields
# 1 passed.

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python pytest -q services/mlx-worker-python/tests/test_model_ops_job_registry.py
# 36 passed.

rm -f coverage.json
PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python coverage run -m pytest -q services/mlx-worker-python/tests/test_model_ops_job_registry.py
PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python coverage json -o coverage.json
python3 scripts/changed_scope_coverage.py --coverage-json coverage.json services/mlx-worker-python/worker/model_ops/job_registry.py services/mlx-worker-python/tests/test_model_ops_job_registry.py
rm -f coverage.json
# TOTAL 6 0 100%.

git diff --cached --check
# passed.

git commit -m "Expose artifact integrity diagnostics summary"
# pre-commit ran make swift-test, make py-test, make integration-test, and scoped performance successfully.
# make swift-test: completed rc=0 elapsed=214.1s.
# make py-test: 4150 passed, 14 skipped, 2 warnings in 213.33s.
# make integration-test: 122 passed, 1 skipped in 458.21s.
# pre-commit scoped performance: status ok; selected probes 2; regressions 0; coverage 100.0%.
# pre-commit performance report: .runtime/pre-commit-performance/20260621-172212-15ceb7de/report/report.md.

git fetch origin main:refs/remotes/origin/main
git rev-parse --short origin/main
git merge-base --is-ancestor origin/main HEAD
# origin/main 946c07eb; origin-main-is-ancestor=no.

git merge --no-edit origin/main
# merged origin/main at 946c07ebbbeffb3bb7a08d4d34bfcf0a77ec639f cleanly.

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python pytest -q services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_managed_download_with_digest_materializes_artifact services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_managed_download_rejects_digest_mismatch_before_activation services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_integrity_rejects_missing_actual_digest_before_activation services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_artifact_integrity_receipt_normalizes_bare_sha256_digest services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_managed_download_rejects_unsupported_digest_policy_before_activation services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_managed_hub_import_with_matching_snapshot_digest_activates services/mlx-worker-python/tests/test_download_pipeline_unit.py::test_strict_managed_hub_import_rejects_snapshot_digest_mismatch_before_activation services/mlx-worker-python/tests/test_managed_artifact_receipts.py::test_strict_managed_hub_download_failure_emits_snapshot_digest_receipt services/mlx-worker-python/tests/test_model_ops_job_registry.py::test_download_registry_snapshot_exposes_operation_receipt_fields
# 9 passed in 0.23s after merging latest origin/main.

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python pytest -q services/mlx-worker-python/tests/test_download_pipeline_unit.py services/mlx-worker-python/tests/test_model_ops_job_registry.py services/mlx-worker-python/tests/test_managed_artifact_receipts.py
# 73 passed in 0.55s after merging latest origin/main.

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python python3 - <<'PY'
from pathlib import Path
from scripts.pre_commit_gate import run_performance_report

changed_files = [
    "docs/plans/2026-06-21-issue-1258-digest-verification.md",
    "docs/plans/2026-06-22-issue-1258-hub-snapshot-digest.md",
    "docs/plans/2026-06-22-issue-1258-diagnostics-integrity-fields.md",
    "docs/runbooks/phase-8-local-install.md",
    "services/mlx-worker-python/tests/test_download_pipeline_unit.py",
    "services/mlx-worker-python/tests/test_managed_artifact_receipts.py",
    "services/mlx-worker-python/tests/test_model_ops_job_registry.py",
    "services/mlx-worker-python/worker/model_ops/download_pipeline.py",
    "services/mlx-worker-python/worker/model_ops/job_registry.py",
]
outcome = run_performance_report(Path.cwd(), changed_files, base_ref="origin/main")
print(f"report_dir={outcome.report_dir}")
print(f"status={outcome.status}")
print(f"selected_probe_count={outcome.selected_probe_count}")
PY
# Status: ok; selected probes 3; regressions 0.
# Report: .runtime/pre-commit-performance/20260621-172516-6be3d91c/report/report.md.
# Coverage: download pipeline 99.0%; job registry 100.0%.
```

## Coverage and Metrics
- Changed-scope coverage: `TOTAL 143 3 98%`.
- Pre-commit scoped coverage: `download_pipeline.py` changed-line coverage 96.47%, `test_download_pipeline_unit.py` changed-line coverage 100%, aggregate 98%.
- Pre-commit performance report: `.runtime/pre-commit-performance/20260621-143628-72ad0e6b/report/report.md`.
- Performance status: `ok`, selected probes 1, regressions 0.
- Probe: `Download pipeline snapshot manifest base reuse`.
- Metrics: `elapsed_ms_mean` base 38.667 ms, head 37.449 ms, delta -3.15%; `elapsed_ms_min` base 38.320 ms, head 36.705 ms, delta -4.21%.
- Post-merge performance replay: `.runtime/pr/perf-merge-check/report/report.md`, status `ok`, selected probes 1, regressions 0; `elapsed_ms_mean` base 38.464 ms, head 37.533 ms, delta -2.42%; `elapsed_ms_min` base 37.442 ms, head 37.045 ms, delta -1.06%.
- Review follow-up changed-scope coverage: `TOTAL 17 0 100%`.
- Review follow-up pre-commit performance report: `.runtime/pre-commit-performance/20260621-154031-6b319861/report/report.md`.
- Review follow-up performance status: `ok`, selected probes 1, regressions 0.
- Review follow-up metrics: `elapsed_ms_mean` base 36.896 ms, head 37.301 ms, delta +1.10% neutral; `elapsed_ms_min` base 36.497 ms, head 36.674 ms, delta +0.49% neutral.
- Latest-origin/main performance replay: `.runtime/pre-commit-performance/20260621-154408-d2017dc4/report/report.md`, status `ok`, selected probes 1, regressions 0, coverage 98.0%.
- Latest-origin/main metrics: `elapsed_ms_mean` base 42.897 ms, head 40.871 ms, delta -4.72% neutral; `elapsed_ms_min` base 39.656 ms, head 40.009 ms, delta +0.89% neutral.
- Hub snapshot digest changed-scope coverage: `TOTAL 128 0 100%`.
- Hub snapshot digest PR-scoped performance report: `.runtime/pr/issue-1258-hub-snapshot/report/report.md`, status `ok`, selected probes 1, regressions 0, coverage 100.0%.
- Hub snapshot digest PR-scoped metrics: `elapsed_ms_mean` base 37.578 ms, head 36.986 ms, delta -1.57% neutral; `elapsed_ms_min` base 36.807 ms, head 36.497 ms, delta -0.84% neutral.
- Hub snapshot digest pre-commit performance report: `.runtime/pre-commit-performance/20260621-162228-d2017dc4/report/report.md`, status `ok`, selected probes 1, regressions 0, coverage 100.0%.
- Hub snapshot digest pre-commit metrics: `elapsed_ms_mean` base 37.198 ms, head 36.711 ms, delta -1.31% neutral; `elapsed_ms_min` base 36.715 ms, head 36.155 ms, delta -1.53% neutral.
- Diagnostics summary changed-scope coverage: `TOTAL 6 0 100%`.
- Diagnostics summary pre-commit performance report: `.runtime/pre-commit-performance/20260621-172212-15ceb7de/report/report.md`, status `ok`, selected probes 2, regressions 0, coverage 100.0%.
- Diagnostics summary pre-commit metrics: derived-model restore `restore_elapsed_ms_mean` base 33.356 ms, head 31.807 ms, delta -4.64% neutral; restore sort `restore_elapsed_ms_mean` base 45.092 ms, head 45.722 ms, delta +1.40% neutral.
- Latest-origin/main aggregate PR-scoped performance report after merging `946c07eb`: `.runtime/pre-commit-performance/20260621-172516-6be3d91c/report/report.md`, status `ok`, selected probes 3, regressions 0.
- Latest-origin/main aggregate coverage: download pipeline 99.0%; job registry 100.0%.
- Latest-origin/main aggregate metrics: download pipeline `elapsed_ms_mean` base 100.382 ms, head 44.821 ms, delta -55.35% improvement; derived-model restore `restore_elapsed_ms_mean` base 37.357 ms, head 33.978 ms, delta -9.04% neutral; restore sort `restore_elapsed_ms_mean` base 47.567 ms, head 46.917 ms, delta -1.37% neutral.
- Observability mode: `evidence`; this slice updates operation receipts only. Probe overhead: `N/A`, no runtime debug/evidence instrumentation added beyond terminal receipt fields.

## Known Gaps
- This is one focused #1258 slice, not full issue closure.
- Deferred: signature verification, release-reference validation, async installer orchestration, desktop UI/schema work, symlink target hashing, and non-regular-file metadata hashing.
- Diagnostics fields are exposed in the Python operation registry snapshot; no desktop UI or protobuf schema changes are included in this slice.
- No protocol or dependency changes; no generated artifacts or lockfiles changed.

## Evidence Checklist
- [x] Relevant plan or spec is identified.
- [x] Behavior changes are reflected in the relevant docs.
- [x] Protocol changes include regenerated generated artifacts, or `N/A` is stated explicitly: no protocol changes.
- [x] Dependency changes include updated lockfiles, or `N/A` is stated explicitly: no dependency changes.
- [x] Relevant tests were run.
- [x] A metrics report is included, or `N/A` is stated explicitly with the reason.
- [x] Observability mode, probe overhead, and any deferred debug or evidence work are recorded, or `N/A` is stated explicitly with the reason.
- [x] Deferred work and known gaps are stated explicitly.

### CODE REVIEW

File: services/mlx-worker-python/worker/model_ops/download_pipeline.py [hunk 1]
```diff
@@ -920,27 +931,139 @@ def uses_operation_receipt(ext: dict[str, str]) -> bool:
...
+            raise ModelOperationError(
+                code="artifact_integrity_unsupported",
+                message="Strict managed artifact installs require a supported SHA-256 digest before activation.",
+                details={
+                    "state_json": state_json,
+                    "failure_reason": failure_reason,
+                },
+            )
+        if not actual_sha256 or declared_sha256 == actual_sha256:
+            return
```
[the remark this answers is not in the document]
Comment (Zigfreidish) [reply services/mlx-worker-python/worker/model_ops/download_pipeline.py]: Fixed in 47e1c095. Strict digest verification now only returns when declared and actual SHA-256 digests match; missing or unparseable actual digests fail closed as artifact_integrity_mismatch before activation. Added regression coverage with test_strict_integrity_rejects_missing_actual_digest_before_activation.

**Review by Zigfreidish — COMMENTED** [review]
## Code Review: Verify managed artifact digests before activation

### Summary
Adds SHA-256 digest verification for strict managed direct downloads — hashes the staged `*.partial` bytes before `os.replace` promotes the artifact. Mismatches and unsupported digest policies fail closed with structured receipts. Clean and correct.

---

### `_sha256_file`

```python
@staticmethod
def _sha256_file(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as file:
        for chunk in iter(lambda: file.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()
```

Streams in 1 MB chunks — correct for large model files, avoids loading the entire artifact into memory.

---

### `_normalize_sha256_digest`

Handles `sha256:<64-hex>` and bare 64-hex formats; lowercases the hex part for comparison stability. Non-SHA-256 strings (e.g. `md5:...`) pass through unchanged so they are recorded as declared but refused as unsupported at the activation gate.

---

### `_sha256_digest_hex`

Returns an empty string for anything that doesn't normalize to a valid `sha256:<64-hex>` digest. Used by `_raise_if_strict_integrity_mismatch` to distinguish "verifiable SHA-256" from "unsupported policy". Clean separation of concerns — the caller never needs to inspect the prefix.

---

### Activation ordering

```python
# 1. hash partial
completed_payload["artifact_integrity"] = self._verified_artifact_integrity_receipt(
    artifact_path=partial_path, ext=..., status="passed"
)
# 2. validate
self._raise_if_strict_integrity_mismatch(
    manifest_payload=completed_payload, partial_path=partial_path, ext=ext
)
# 3. activate (only if 2 did not raise)
os.replace(os.fspath(partial_path), os.fspath(output_path))
```

Fail-closed: the partial is never promoted on mismatch. On failure, the partial is left intact with `partial_lifecycle="failed_kept_for_resume"` and `activated=False`.

---

### Error codes

Two distinct codes for two distinct failure modes:
- `artifact_integrity_mismatch` — bytes don't match declared SHA-256.
- `artifact_integrity_unsupported` — declared policy exists but isn't a verifiable SHA-256 format.

This allows callers to distinguish byte corruption from policy configuration errors.

---

### `test_strict_integrity_rejects_missing_actual_digest_before_activation`

Tests a receipt that claims `status=passed` but has an empty `actual_digest`. The method correctly refuses activation — this closes a potential bypass where a pre-built receipt with no computed hash could slip through without actually running `_sha256_file`.

---

### `_artifact_integrity_receipt` refactor

Changed from `@staticmethod` to `@classmethod` to call `cls._artifact_digest` and `cls._normalize_sha256_digest` without repeating the normalization logic. `actual_digest` is only added to the receipt when `digest` is non-empty, keeping the schema minimal for non-digest requests.

`verification_mode` is promoted to `"sha256"` when either declared or actual digest normalizes as SHA-256 — correct: the mode reflects what was actually verified, not what was declared.

---

### Minor Observations

- `checked_at` uses `time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())` — UTC ISO-8601 without microseconds, consistent with the existing receipt format.
- `_build_manifest_payload` path now builds `completed_payload` first and amends `artifact_integrity` before writing the manifest JSON. Previously, the snapshot was appended after `os.replace` with a separate `_snapshot()` call. The new path correctly defers the manifest write until after digest verification.

---

### Verdict
Correct fail-closed implementation, clear error codes, well-ordered operations, comprehensive edge-case coverage. 4144 Python + 122 integration tests passed after merging origin/main. Would approve if self-approval were permitted.

---
_Generated by [Claude Code](https://claude.ai/code)_
